store/tikv/2pc_utils.go

```diff
@@ -0,0 +1,397 @@
+	it.keep = f
+}
+
+type staticMutations struct {
+	mutations CommitterMutations
+}
+
+func (m staticMutations) Iter(start, end []byte) mutationsIter {
+	it := &staticMutationsIter{
+		mutations: m.mutations,
+		end:       end,
+	}
+	if len(start) != 0 {
+		it.idx = sort.Search(len(m.mutations.keys), func(i int) bool {
+			return bytes.Compare(m.mutations.keys[i], start) >= 0
+		})
+	}
+	return it
+}
+
+func (m staticMutations) Len() int {
+	return m.mutations.len()
+}
+
+type staticMutationsIter struct {
+	mutations CommitterMutations
+	idx       int
+	end       []byte
+	keep      func([]byte) bool
+}
+
+func (it *staticMutationsIter) Next() mutation {
+	for ; it.idx < len(it.mutations.keys); it.idx++ {
+		m := mutation{
+			key: it.mutations.keys[it.idx],
+		}
+		if it.idx < len(it.mutations.values) {
+			m.value = it.mutations.values[it.idx]
+			m.op = it.mutations.ops[it.idx]
+			m.isPessimisticLock = it.mutations.isPessimisticLock[it.idx]
+		}
+		if it.keep != nil && !it.keep(m.key) {
+			continue
+		}
+
+		it.idx++
+		return m
+	}
+	return mutation{}
+}
+
+func (it *staticMutationsIter) WithFilter(f func([]byte) bool) {
+	it.keep = f
+}
+
+type committerTxnMutations struct {
+	*twoPhaseCommitter
+	isPrewrite bool
+}
+
+func (m committerTxnMutations) Iter(start, end []byte) mutationsIter {
+	return m.newMutationsIter(m.txn.GetMemBuffer().IterWithFlags(start, end), m.isPrewrite)
+}
+
+func (m committerTxnMutations) Len() int {
+	base := m.txn.Len() - m.ignoredKeys
+	if m.isPrewrite {
+		return base
+	}
+	return base - m.prewriteOnlyKeys
+}
+
+type txnMutationsIter struct {
+	src           kv.MemBufferIterator
+	isPrewrite    bool
+	isPessimistic bool
+	keep          func(key []byte) bool
+
+	prewriteOnlyKeys int
+	ignoredKeys      int
+}
+
+func (c *twoPhaseCommitter) newMutationsIter(src kv.MemBufferIterator, isPrewrite bool) *txnMutationsIter {
+	return &txnMutationsIter{
+		src:           src,
+		isPrewrite:    isPrewrite,
+		isPessimistic: c.txn.IsPessimistic(),
+	}
+}
+
+func (it *txnMutationsIter) WithFilter(f func(key []byte) bool) {
+	it.keep = f
+}
+
+func (it *txnMutationsIter) Next() (m mutation) {
+	var err error
+	for src := it.src; src.Valid(); err = src.Next() {
+		_ = err
+		m.key = src.Key()
+		m.value = nil
+		m.isPessimisticLock = false
+		flags := src.Flags()
+
+		if it.keep != nil && !it.keep(m.key) {
+			continue
+		}
+
+		if flags.HasIgnoredIn2PC() {
+			continue
+		}
+
+		if it.isPrewrite {
+			ignored := it.fillMutationForPrewrite(&m, src)
+			if ignored {
+				src.UpdateFlags(kv.SetIgnoredIn2PC)
+				it.ignoredKeys++
+				continue
+			}
+		} else {
+			// For commit and cleanup, we only need keys.
+			if flags.HasPrewriteOnly() {
+				continue
+			}
+		}
+
+		err = src.Next()
+		_ = err
+		return
+	}
+	return mutation{}
+}
+
+func (it *txnMutationsIter) fillMutationForPrewrite(m *mutation, src kv.MemBufferIterator) bool {
+	flags := src.Flags()
+
+	if !src.HasValue() {
```

Comment:
It's confusing that the entry HasValue is false, then at line 370 check `len(value) > 0`.
If `len(value) == 0` is not the same as `!HasValue`, can we make it more clear?

Reply:
After `MemBuffer.DiscardValues` we cannot get `value`.

Reply:
Do you mean the `src.HasValue` is for the whole MemBuffer, but `src.Value` returns the current value?

Reply:
The implementation just check the current entry.
My point is, `HasValue` semantically should be the same as `len(src.Value()) > 0` but actually is not.

Reply:
No, flag only keys do not have `Value`, and `len(src.Value()) == 0` is true for tombstone.
`HasValue` is used to check whether the current key is flags only, `len(...) > 0` is for key deletion.

Reply:
`src.HasValue() && len(sec.Value()) == 0` will be true for tombstones.

Reply:
The logical is, we can `GetValue` if and only if current key `HasValue`